### PR TITLE
Ensure discards are initially soft selected in VSCode

### PR DESCRIPTION
**Pull request by dibarbet**
Resolves https://github.com/dotnet/vscode-csharp/issues/7623

Unfortunately LSP doesn't give servers much control over filtering.  Servers do not generally get callbacks as filter text is typed - meaning we generally are not able to modify hard vs soft selection after the initial list is presented.

However, we can force VSCode to call us back in some scenarios using the `isIncomplete` flag on the completion list.  For this case, when we see only punctuation typed (`_`), we set the `isIncomplete` flag and unset the commit characters to mimic soft selection.

As the user continues typing (e.g. `_o`) we get called again and we can update to hard selection if they are completing a variable.  

However - this does not cover every case of switching between hard/soft selection.  It could be covered by always setting `isIncomplete` but that would be expensive as we're re-computing and reserializing the entire list.

### CODE REVIEW

File: src/Features/Core/Portable/Completion/CompletionService.cs [hunk 1]
```diff
@@ -363,6 +363,82 @@ internal static void FilterItems(
         }
     }
 
+    internal static bool IsHardSelection(CompletionItem item, bool matchedFilterText, bool hasSuggestionModeItem, string filterText)
```
Comment (dibarbet) [inline src/Features/Core/Portable/Completion/CompletionService.cs:366]: everything in this file is moved unchanged

File: src/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs [hunk 2]
```diff
@@ -253,5 +278,67 @@ static CompletionFilterReason GetFilterReason(CompletionTrigger trigger)
                 };
             }
         }
+
+        private static MatchResult GetBestCompletionItemSelectionFromFilteredResults(IReadOnlyList<MatchResult> filteredMatchResults)
```
Comment (genlu) [inline src/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:282]: is this copied from ItemManager? We uses this to actually determine the selected item in VS, but here this is at best a heuristic which might not match VSCode selection.

File: src/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs [hunk 3]
```diff
@@ -219,6 +219,29 @@ private static (CompletionList CompletionList, bool IsIncomplete) FilterCompleti
             // Next, we sort the list based on the pattern matching result.
             matchResultsBuilder.Sort(MatchResult.SortingComparer);
 
+            // Determine if the list should be hard selected or soft selected.
+            bool isHardSelection;
+            if (matchResultsBuilder.Count > 0)
+            {
+                var bestResult = GetBestCompletionItemSelectionFromFilteredResults(matchResultsBuilder);
+                isHardSelection = CompletionService.IsHardSelection(bestResult.CompletionItem, bestResult.ShouldBeConsideredMatchingFilterText, completionList.SuggestionModeItem != null, filterText);
```
Comment (genlu) [inline src/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:227]: is this part needed to handle soft-selecting `_`? I think we could just use the result of `IsAllPunctuation(filterText)` to decide both selection mode and isIncomplete

File: src/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs (hunk 2 above)
Comment (dibarbet) [reply src/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:282]: Yes this is copied.  Understood that it is a heuristic which may not match, but this definitely works better than before

File: src/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs (hunk 3 above)
Comment (dibarbet) [reply src/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:227]: Ah interesting, I can make that change

File: src/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs (hunk 2 above)
Comment (dibarbet) [reply src/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs]: reverted

File: src/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs [hunk 4]
```diff
@@ -219,6 +219,21 @@ private static (CompletionList CompletionList, bool IsIncomplete) FilterCompleti
...
+            // If they type something that is not punctuation, we may need to update the hard vs soft selection.
+            // For example, typing '_' should initially be soft selection, but if the user types 'o' we should hard select '_otherVar' (if it exists).
+            // This isn't perfect - ideally we would make this determination every time a filter character is typed, but we do not get called back
+            // for typing filter characters in LSP (unless we always set isIncomplete, which is expensive).
+            var isHardSelection = completionList.SuggestionModeItem is null && !isFilterTextAllPunctuation;
+            var isIncomplete = isFilterTextAllPunctuation;
+
+            // If our completion list hasn't hit the max size, we don't need to do anything filtering
+            if (completionListMaxSize < 0 || completionListMaxSize >= completionList.ItemsList.Count)
+                return (completionList, isIncomplete, isHardSelection);
```
Comment (genlu) [inline src/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:235]: I think this part can be moved to the beginning of the method, before calculating MatchResults

Comment (dibarbet) [reply src/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:235]: ah yup now that it no longer needs it.  moved

**Review by genlu — APPROVED** [review]

**Pull request merged**